src/saltext/vmware/modules/roles.py

```diff
@@ -0,0 +1,337 @@
+        for privilege in role["privileges"]:
+            priv_name = privilege["name"]
+            group_name = privilege["groupLabel"]
+            if group_name not in role_json["privileges"]:
+                role_json["privileges"][group_name] = []
+            role_json["privileges"][group_name].append(priv_name)
+        roles_config.append(role_json)
+
+    return roles_config
+
+
+def save(role_config, service_instance=None, profile=None):
+    """
+    Create new role with given configuration, if it doesn't exist.
+    Otherwise update existing role.
+    Apply changes only for particular group from configuration.
+    Roles outside the groups mentioned in configuration are kept unchanged.
+
+    role_config
+        Role name and configuration values.
+
+    service_instance
+        Use this vCenter service connection instance instead of creating a new one. (optional).
+
+    profile
+        Profile to use (optional)
+
+    .. code-block:: json
+
+        {
+            "role": "SRM Administrator",
+            "privileges": {
+                "Protection Group": [
+                    "Assign to plan",
+                    "Create",
+                    "Modify",
+                    "Remove",
+                    "Remove from plan"
+                ],
+                "Recovery Plan": [
+                    "Configure commands",
+                    "Create",
+                    "Remove",
+                    "Modify",
+                    "Recovery"
+                ]
+            }
+        }
+    """
+    service_instance = service_instance or connect.get_service_instance(
+        config=__opts__, profile=profile
+    )
+
+    authorizationManager = service_instance.RetrieveContent().authorizationManager
+
+    privileges_desc = _get_privilege_descriptions(authorizationManager)
+    privilege_groups_desc = _get_privilege_group_descriptions(authorizationManager)
+
+    # Collect privilages by group label and privilege name
+    group_privileges = {}
+    privilege_group_map = {}
+    for privilege in authorizationManager.privilegeList:
+        desciption_group = privilege_groups_desc[privilege.privGroupName]
+        group_label = desciption_group["label"]
+        if group_label not in group_privileges:
+            group_privileges[group_label] = []
+        group_privileges[group_label].append(privilege)
+        privilege_group_map[privilege.privId] = group_label
+
+    role_name = role_config["role"]
+
+    # find role to store or update
+    role = None
+    for role_obj in authorizationManager.roleList:
+        if role_obj.info.label == role_name:
+            role = role_obj
+            break
+
+    # group new privileges
+    new_privileges_by_groups = {}
+    for group in role_config["privileges"]:
+        if group not in new_privileges_by_groups:
+            new_privileges_by_groups[group] = []
+        privileges_in_group = role_config["privileges"][group]
+        for priv in group_privileges[group]:
```

Comment:
**suggestion** try this with a non-existent group, e.g. `salt-call vcenter_roles.save '{"role": "Blerps", "privileges": {"Fnord": ["Create resource pool",  "something else", "cool guy"]}}'`

You should see this raise a `KeyError`, which is not so good :sweat_smile: